### ISSUE TITLE
build_hist is (heavily) latency bounded

### ISSUE THREAD

**Issue by Laurae2** [hcho3/xgboost-fast-hist-perf-lab#12]
After the enhancements of #11, I am noticing the build_hist is heavily latency bounded when running in parallel.

Specifically, this line just eats all the CPU time in parallel: https://github.com/hcho3/xgboost-fast-hist-perf-lab/blob/master/src/build_hist.cc#L12

Multiple solutions from there:

* Find whether there is false sharing occurring (done), and fix it with padding, how?
* Get rid of the memory allocation by doing something else, how?
* Get rid of the data variable by writing directly to the hist, how?
* Rewrite the algorithm, what?

According to VTune:

![image](https://user-images.githubusercontent.com/9083669/48220369-60e13300-e38f-11e8-8737-217fae25b9a2.png)

![image](https://user-images.githubusercontent.com/9083669/48220445-9554ef00-e38f-11e8-9f03-07a2651f81b7.png)

In general:

![image](https://user-images.githubusercontent.com/9083669/48221345-1ad99e80-e392-11e8-8388-ca10afe070da.png)

The memory latency issue on the data allocation:

![image](https://user-images.githubusercontent.com/9083669/48221413-51afb480-e392-11e8-8196-3a2ccaa92c02.png)

**Comment by Laurae2** [comment]
With a loop instead of fill.

gcc:

![image](https://user-images.githubusercontent.com/9083669/48284702-6f4b4f80-e460-11e8-8453-c5b5f8088695.png)

![image](https://user-images.githubusercontent.com/9083669/48284755-a91c5600-e460-11e8-8b9e-da4fa2b01678.png)

icc:

![image](https://user-images.githubusercontent.com/9083669/48285721-c9014900-e463-11e8-9cc1-09e709f75c10.png)

![image](https://user-images.githubusercontent.com/9083669/48285645-8f304280-e463-11e8-8d4c-baa0eac272de.png)

**Comment by Laurae2** [comment]
Found a workaround, new issue incoming to describe the workaround. But I had to ignore the private BuildHist variable private (public to main thread) though.
